pyiceberg/io/pyarrow.py

```diff
@@ -638,10 +634,124 @@ def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> p
+        self._handle_explicit_is_null_or_not(term)
+
+    def visit_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_not_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_greater_than_or_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_greater_than(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_less_than(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_less_than_or_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_starts_with(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_not_starts_with(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_skipped(term)
+
+    def visit_true(self) -> None:
+        return
+
+    def visit_false(self) -> None:
+        return
+
+    def visit_not(self, child_result: pc.Expression) -> None:
+        return
+
+    def visit_and(self, left_result: pc.Expression, right_result: pc.Expression) -> None:
+        return
+
+    def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> None:
+        return
+
+
+def _get_is_valid_or_not_bound_refs(expr: BooleanExpression) -> tuple[Set[BoundReference[Any]], Set[BoundReference[Any]]]:
+    """Collect the bound terms catogorized by having at least one is_null or is_not_null in the expr and the remaining."""
+    collector = _CollectNullUnmentionedTermsFromExpression()
+    boolean_expression_visit(expr, collector)
+    null_unmentioned_bound_terms = collector.null_unmentioned_bound_terms
+    is_valid_or_not_bound_terms = collector.is_valid_or_not_bound_terms
+
+    null_unmentioned_bound_refs: Set[BoundReference[Any]] = set()
+    is_valid_or_not_bound_refs: Set[BoundReference[Any]] = set()
+    for t in null_unmentioned_bound_terms:
+        if not isinstance(t, BoundReference):
+            raise ValueError("Collected Bound Term that is not reference.")
+        else:
+            null_unmentioned_bound_refs.add(t)
+    for t in is_valid_or_not_bound_terms:
+        if not isinstance(t, BoundReference):
+            raise ValueError("Collected Bound Term that is not reference.")
+        else:
+            is_valid_or_not_bound_refs.add(t)
+    return null_unmentioned_bound_refs, is_valid_or_not_bound_refs
+
+
 def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
     return boolean_expression_visit(expr, _ConvertToArrowExpression())
 
 
+def expression_to_reverted_pyarrow(expr: BooleanExpression) -> pc.Expression:
+    """Complimentary filter convertion function of expression_to_pyarrow.
```

Comment:
Could we mention that this function is used specifically for reverting an expression to pyarrow expression for the delete operation? As we discussed offline, negating a pc.Expression has the expected outcome that ignores Nulls, and only becomes a problem when we use it for deletion, because we want to avoid deleting Nulls unless it is specified in the delete_filter explicitly. On that note, I feel that it may be good to make this a private function as well. WDYT?

nit: typo 
```suggestion
    """Complimentary filter conversion function of expression_to_pyarrow.
```

Reply:
I think your suggestion totally makes sense but I am on the fence. I think the semantics is a little broader and might be useful to user? because we have 1.raw expr, then 2.'revert' expression (by ~pc_expr or by expression_to_pyarrow(Not(iceberg_expr))) and then 3.this expr which is 'comlimentary'. what do you think if we change the name from reverted to complimentary?